test/internals/request.test.js

```diff
@@ -266,3 +273,470 @@ test('Request with trust proxy and undefined socket', t => {
+      const second = req.validate(defaultSchema, null, 'querystring')
+
+      t.equal(first, second)
+      t.ok(first({ hello: 'world' }))
+      t.ok(second({ hello: 'world' }))
+      t.equal(called, 2)
+
+      reply.send({ hello: 'world' })
+    })
+
+    await fastify.inject({
+      path: '/',
+      method: 'GET'
+    })
+  })
+
+  subtest.test('Should return true/false if input valid - With Schema for Route defined', async t => {
+    const fastify = Fastify()
+    let validate = null
+
+    t.plan(9)
+
+    fastify.post('/:id', {
+      schema: requestSchema
+    }, (req, reply) => {
+      const { params } = req
+
+      switch (params.id) {
+        case 1:
+          validate = req.validate(null, null, 'body')
+          t.ok(validate({ hello: 'world' }))
+          t.notOk(validate({ hello: [], world: 'foo' }))
+          break
+        case 2:
+          t.notOk(req.validate(null, { foo: 'something' }, 'querystring'))
+          t.ok(req.validate(null, { foo: 'bar' }, 'querystring'))
+          break
+        case 3:
+          t.notOk(req.validate(null, { 'x-foo': [] }, 'headers'))
+          t.ok(req.validate(null, { 'x-foo': 'something' }, 'headers'))
+          break
+        case 4:
+          t.ok(req.validate(null, { id: params.id }, 'params'))
+          t.notOk(req.validate(null, { id: 0 }, 'params'))
+          break
+        case 5:
+          t.equal(validate, req.validate(null, null, 'body'))
+          break
+        default:
+          t.fail('Invalid id')
+      }
+
+      reply.send({ hello: 'world' })
+    })
+
+    const promises = []
+
+    for (let i = 1; i < 6; i++) {
+      promises.push(fastify.inject({
+        path: `/${i}`,
+        method: 'post',
+        query: { foo: 'bar' },
+        payload: {
+          hello: 'world'
+        },
+        headers: {
+          'x-foo': 'x-bar'
+        }
+      }))
+    }
+
+    await Promise.all(promises)
+  })
+})
+
+test('Request#serialize', subtest => {
+  const defaultSchema = {
+    type: 'object',
+    required: ['hello'],
+    properties: {
+      hello: { type: 'string' },
+      world: { type: 'string' }
+    }
+  }
+  const responseSchema = {
+    201: {
+      type: 'object',
+      properties: {
+        status: {
+          type: 'string',
+          enum: ['ok']
+        },
+        message: {
+          type: 'string'
+        }
+      }
+    },
+    '4xx': {
+      type: 'object',
+      properties: {
+        status: {
+          type: 'string',
+          enum: ['error']
+        },
+        code: {
+          type: 'integer',
+          minimum: 1
+        },
+        message: {
+          type: 'string'
+        }
+      }
+    }
+  }
+
+  subtest.plan(5)
+
+  subtest.test('Should return a function when empty input - Route without schema', async t => {
+    const fastify = Fastify()
+
+    t.plan(4)
+
+    fastify.get('/', (req, reply) => {
+      const serialize = req.serialize(defaultSchema)
```

Comment:
I'm thinking it is confusing to me that the `serialize` function is attached to the `request` object: all the serialize stuff is used during the response flow, so I would see it better to the reply object actually

Reply:
I actually wasn’t sure as well if request was the best place, but I agree that reply seems more appropriate as it’s its responsibility to prepare the response ahead of dispatching it. I’ll make the proper changes